### PR TITLE
Added libsodium, created an abstracted encrypted storage provider, cr…

### PR DESCRIPTION
**Pull request by chancity**
Created new local storage provider using abstracted encrypted storage provider, created a remote storage provider implementing the new encrypted storage provider.

The DefaultKeyStoreCryptoProvider is compatible with the current android and iOS backup keystore format (json properties, salt, encryption, etc).  This can be leveraged to import QRCODE backups from a different application within the Kin Ecosystem.

The abstracted storage provider ensures that private seeds are never accessible via JavaScript object.  

The KinSdk.KeystoreProvider sign implementation is current not compatible with this different implementation because a passphrase is required for any signing of a transaction envelope.  I also didn't want to presist the secret passphrase required for decryption. 

I currently can't think of a good way to not require a secret passphrase for any interaction, maybe a callback called in the sign implementation?

### CODE REVIEW

**Comment by chaseeb** [comment]
@katzio would you mind taking a look at this and see if it can be helpful =)

**Comment by chancity** [comment]
@katzio hi

File: scripts/src/EncryptedStorageProviders/utils.ts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+const KIN_WALLET_STORAGE_INDEX = 'kin-wallet';
+export const getIndex = (id?: string) => `${KIN_WALLET_STORAGE_INDEX}${id && `_${id}`}`;
```
Comment (katzio) [inline scripts/src/EncryptedStorageProviders/utils.ts]: Just verifying my thoughts: this opens the option for *multiple accounts*?

**Comment by katzio** [comment]
Nice! 
Correct me if I'm wrong, this is an enhanced version of the already existing encryption method we implemented right? 
so [localstorage-provider.ts](https://github.com/kinecosystem/kin-sdk-js-keystore-providers/blob/master/scripts/src/localstorage-provider.ts) and [local-storage-handler.ts](https://github.com/kinecosystem/kin-sdk-js-keystore-providers/blob/master/scripts/src/libs/local-storage-handler.ts) are redundant 


I'll try to fix the conflicts and see how we merge it with what we did already.
Thanks! 

Some notes

> The abstracted storage provider ensures that private seeds are never accessible via JavaScript object.

They are stored in localStorage, anything can access them (encrypted but...)
Or you mean they are out of scope while in memory?

> I currently can't think of a good way to not require a secret passphrase for any interaction, maybe a callback called in the sign implementation?

We thought about using some kind of **timed session** with the KeystoreProvider providing (in memory, but without public access) the **decrypted pk** so the user can interact without entering its passphrase over and over again.


**Edit:** conflicts resolved. 
Looks like you did a way better job with the crypto implementation, I'll refactor the `localstorage-provider.ts` so it won't use crypto and merge yours.

**Comment by katzio** [comment]
If you could figure out why Webpack bundles the whole KinSdk into the Encrypted Providers would be much appreciated

```
./encrypted-localstorage-provider.js   857 KiB                 0  [emitted]  [big]  encrypted-localstorage-provider
./encrypted-remotestorage-provider.js   858 KiB                 1  [emitted]  [big]  encrypted-remotestorage-provider
```

File: scripts/src/EncryptedStorageProviders/utils.ts (hunk 1 above)
Comment (chancity) [reply scripts/src/EncryptedStorageProviders/utils.ts]: @katzio it does if an ID is passed

**Comment by chancity** [comment]
> Nice!
> Correct me if I'm wrong, this is an enhanced version of the already existing encryption method we implemented right?
> so [localstorage-provider.ts](https://github.com/kinecosystem/kin-sdk-js-keystore-providers/blob/master/scripts/src/localstorage-provider.ts) and [local-storage-handler.ts](https://github.com/kinecosystem/kin-sdk-js-keystore-providers/blob/master/scripts/src/libs/local-storage-handler.ts) are redundant
> 
> I'll try to fix the conflicts and see how we merge it with what we did already.
> Thanks!
> 
> Some notes
> 
> > The abstracted storage provider ensures that private seeds are never accessible via JavaScript object.
> 
> They are stored in localStorage, anything can access them (encrypted but...)
> Or you mean they are out of scope while in memory?
> 
> > I currently can't think of a good way to not require a secret passphrase for any interaction, maybe a callback called in the sign implementation?
> 
> We thought about using some kind of **timed session** with the KeystoreProvider providing (in memory, but without public access) the **decrypted pk** so the user can interact without entering its passphrase over and over again.
> 
> **Edit:** conflicts resolved.
> Looks like you did a way better job with the crypto implementation, I'll refactor the `localstorage-provider.ts` so it won't use crypto and merge yours.

The crypto implementation should be the same as the android and ios backup implementations.

**Comment by chancity** [comment]
@katzio The bundle is not including the KinSdk, it is including libsodium and libsodium-wrappers https://www.npmjs.com/package/libsodium-wrappers-sumo.  Can either make it a dev dependency and have developers reference their own version or leave it.  What do you think is best?

**Pull request merged**